lib/CL/pocl_cache.c

```diff
@@ -0,0 +1,470 @@
+    assert(bytes_written > 0 && bytes_written < POCL_FILENAME_LENGTH);
+
+    return pocl_write_file(descriptor, content, size, 0, 1);
+}
+
+/******************************************************************************/
+
+
+char* pocl_cache_read_buildlog(cl_program program,
+                               unsigned device_i) {
+    char buildlog_path[POCL_FILENAME_LENGTH];
+    program_device_dir(buildlog_path, program,
+                       device_i, POCL_BUILDLOG_FILENAME);
+
+    char* res=NULL;
+    uint64_t filesize;
+    if (pocl_read_file(buildlog_path, &res, &filesize))
+        return NULL;
+
+    res[filesize]='0';
+    return res;
+}
+
+
+int pocl_cache_append_to_buildlog(cl_program  program,
+                                  unsigned    device_i,
+                                  const char *content,
+                                  size_t      size) {
+    char buildlog_path[POCL_FILENAME_LENGTH];
+    program_device_dir(buildlog_path, program,
+                       device_i, POCL_BUILDLOG_FILENAME);
+
+    return pocl_write_file(buildlog_path, content, size, 1, 1);
+}
+
+/******************************************************************************/
+
+static int make_kernel_cachedir_path(char*        kernel_cachedir_path,
+                                     cl_program   program,
+                                     unsigned     device_i,
+                                     cl_kernel    kernel,
+                                     size_t       local_x,
+                                     size_t       local_y,
+                                     size_t       local_z) {
+    assert(kernel->name);
+    char tempstring[POCL_FILENAME_LENGTH];
+
+    int bytes_written = snprintf(tempstring, POCL_FILENAME_LENGTH,
+                                 "/%s/%zu-%zu-%zu", kernel->name,
+                                 local_x, local_y, local_z);
+    assert(bytes_written > 0 && bytes_written < POCL_FILENAME_LENGTH);
+
+    program_device_dir(kernel_cachedir_path, program, device_i, tempstring);
+
+    return pocl_mkdir_p(kernel_cachedir_path);
+}
+
+
+int pocl_cache_write_kernel_parallel_bc(void*        bc,
+                                        cl_program   program,
+                                        unsigned     device_i,
+                                        cl_kernel    kernel,
+                                        size_t       local_x,
+                                        size_t       local_y,
+                                        size_t       local_z) {
+    assert(bc);
+
+    char kernel_parallel_path[POCL_FILENAME_LENGTH];
+    make_kernel_cachedir_path(kernel_parallel_path, program, device_i,
+                              kernel, local_x, local_y, local_z);
+
+    assert( strlen(kernel_parallel_path) <
+            (POCL_FILENAME_LENGTH - strlen(POCL_PARALLEL_BC_FILENAME)));
+    strcat(kernel_parallel_path, "/" POCL_PARALLEL_BC_FILENAME);
```

Comment:
Nope, not in POCL_PARALLEL_BC_FILENAME. That define is still in pocl_cl.h because it's used in other places (like devices/common.c). But i can move it to pocl_cache and add the "/", it would be more consistent, i guess.